### ISSUE TITLE
[DOC] [DISCUSS] document supported joins, preferably using annotations

### ISSUE THREAD

**Issue by abellina** [NVIDIA/spark-rapids#343]
In general terms, we implement Spark execs (like join) in the gpu, and there could be caveats. Say at some point we don't support `RightJoin`, or at some point input with `NaN`'s is not supported. It seems this can be easily handled using an annotation at the exec level that describes (in simple terms) what the caveat should be, so we can pick it up into `configs.md` or some other page.

Here's an example:

```
@GpuExpressionDescrition(
   gpuCaveat="doesn't implement feature X")
case class MyExec()
```

Then we can call `getAnnotation` in the class to get this data back, like the `FunctionRegistry` does.

Perhaps we can also move the override description to this annotation as well, and add a `gpuSupportedType` property to enumerate supported types.

**Comment by revans2** [comment]
I am just curious how much effort we want to put into this.  If it is just a string, then we can include it in the description we already have implemented.  Or perhaps a separate field like we do for incompat.

https://github.com/NVIDIA/spark-rapids/blob/468abaa328d45e330ea4a7f7918201784b5a82e3/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L1724

**Comment by revans2** [comment]
Sorry one more thing.  The problem with all of this is trying to keep everything in sync, like any other documentation.  If we have to keep a text string somewhere in sync with the code it describes, we want to be sure that the code and the string are as close to each other as possible, and even then they will get out of sync.  The best would be to do something with the actual code.  If we could make some of the checks data driven, like input types, and scalar vs non-scalar parameters, then we could have both the documentation and the checks run off of the same code.  That to me is what we should be shooting for.

Annotations don't appear to solve any of that for me.

**Comment by abellina** [comment]
> The best would be to do something with the actual code. If we could make some of the checks data driven, like input types, and scalar vs non-scalar parameters, then we could have both the documentation and the checks run off of the same code. That to me is what we should be shooting for.

I agree, but this goes back to the level of effort. I think for types, the effort is worth it though, and I can see ways of tackling it. It seems to me that given a set of `DataType` we should be able to call a static function that returns some kind of _SupportLevel_ instance that can state that level: "supported", "not supported", "supported IF condition", and the condition expressed in a way that can be evaluated (hands waving).

> Annotations don't appear to solve any of that for me.

Well they do bring the info closer to the [code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala#L26), but yeah no guarantee that this would be updated => it's not executable. That said I do not know what the alternative is, because there could be things not easily captured (or not possible to capture). I can think of ways to express that the gpu join only supports certain `JoinType` instances, something like:

```
  validIf(join.joinType)
      .matches(Inner, Cross)
      .otherwise(s"$join.joinType currently is not supported")
```

But then you have the cases where we depend on something at runtime:

```
    val left = childPlans.head.convertIfNeeded()
    val right = childPlans(1).convertIfNeeded()
    // The broadcast part of this must be a BroadcastExchangeExec
    val buildSide = join.buildSide match {
      case BuildLeft => left
      case BuildRight => right
    }
    if (!buildSide.isInstanceOf[GpuBroadcastExchangeExec]) {
      throw new IllegalStateException("the broadcast must be on the GPU too")
    }
```

And yeah this one we can't really capture in a good way. Unless I am missing something.
